azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,405 @@
+        The problem is any extensions contained here may be out of date, because they were
+        more recently updated via a FastTrack GS. Therefore, we remove them here in that case.
+        """
+        if self._pending_mode == GOAL_STATE_SOURCE_FABRIC:
+            svd_seq_no = self._last_svd_seq_no
+            if svd_seq_no is None:
+                svd_seq_no = self._get_saved_svd_seqNo()
+            if str(extensions_config.svd_seqNo) == str(svd_seq_no):
+                logger.info("SvdSeqNo did not change. Removing extensions from goal state")
+                extensions_config.ext_handlers = None
+                self._last_svd_seq_no = extensions_config.svd_seqNo
+                self._pending_svd_seq_no = extensions_config.svd_seqNo
+            else:
+                self._pending_svd_seq_no = svd_seq_no
+
+    def _decide_what_to_process(self, fabric_changed, fast_track_changed):
+        """
+        If just Fabric GS changed, then process only that.
+        If just FastTrack GS changed, then process only that.
+        If both changed, then process Fabric and then FastTrack.
+        If neither changed, then process whichever we used last (to keep with the current behavior)
+        """
+        if fabric_changed:
+            self._set_reason(ExtensionsConfigReasons.FABRIC_CHANGED)
+            return GOAL_STATE_SOURCE_FABRIC
+        if fast_track_changed:
+            self._set_reason(ExtensionsConfigReasons.FAST_TRACK_CHANGED)
+            return GOAL_STATE_SOURCE_FASTTRACK
+
+        mode = self._last_mode
+        if mode is None:
+            mode = self._get_saved_mode()
+        if mode == GOAL_STATE_SOURCE_FASTTRACK:
+            self._set_reason(ExtensionsConfigReasons.FAST_TRACK_LAST_CHANGE)
+            return GOAL_STATE_SOURCE_FASTTRACK
+
+        self._set_reason(ExtensionsConfigReasons.FABRIC_LAST_CHANGE)
+        return GOAL_STATE_SOURCE_FABRIC
+
+    def _set_reason(self, reason):
+        self._reason = "{0} FastTrack={1}, Fabric={2}".format(reason, self._fast_track_changed_detail, self._fabric_changed_detail)
+
+    def _get_fast_track_changed(self, artifacts_profile):
+        if artifacts_profile is None:
+            self._fast_track_changed_detail = FastTrackChangeDetail.NO_PROFILE
+            return False
+        if not artifacts_profile.has_extensions():
+            self._fast_track_changed_detail = FastTrackChangeDetail.NO_EXTENSIONS
+            return False
+
+        sequence_number = self._get_last_sequence_number()
+        if sequence_number is None or sequence_number != artifacts_profile.get_sequence_number():
+            self._fast_track_changed_detail = FastTrackChangeDetail.SEQ_NO_CHANGED
+            return True
+
+        self._fast_track_changed_detail = FastTrackChangeDetail.NO_CHANGE
+        return False
+
+    def _get_fabric_changed(self, goal_state_incarnation):
+        if goal_state_incarnation is None:
```

Comment:
Under what conditions can the incarnation be None?

Reply:
With the latest changes, only at startup